keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1615 @@
+notably, the ones using [`TestUnderTemporaryNetworkFailure`].407466
+
+[It should recreate its iptables rules if they are deleted]: https://github.com/kubernetes/kubernetes/blob/v1.27.0/test/e2e/network/networking.go#L550
+[`TestUnderTemporaryNetworkFailure`]: https://github.com/kubernetes/kubernetes/blob/v1.27.0-alpha.2/test/e2e/framework/network/utils.go#L1078
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+- <test>: <link to test coverage>
+
+#### Scalability tests
+
+TBD
+
+### Graduation Criteria
+
+#### Alpha
+
+- `kube-proxy --proxy-mode nftables` available behind a feature gate
+
+- nftables mode has unit test parity with iptables
+
+- An nftables-mode e2e job exists, and passes
+
+#### Beta
+
+- At least two releases since Alpha.
+
+- The nftables mode has seen real-world usage, no major bugs.
+
+- A "kube-proxy mode-switching" e2e job exists, to confirm that you
+  can redeploy kube-proxy in a different mode in an existing cluster.
+  (eg, it tests "`kube-proxy --cleanup`")
+
+- An nftables e2e periodic perf/scale job exists, and shows
+  performance as good as iptables and ipvs.
+
+- No UNRESOLVED sections in the KEP. (In particular, we have figured
+  out what sort of "API" we will offer for integrating third-party
+  nftables rules.)
+
+#### GA
+
+- At least two releases since Beta.
+
+- The nftables mode has no bugs / regressions that would make us
+  hesitate to recommend it.
+
+- We have at least the start of a plan for the next steps (changing
+  the default mode, deprecating the old backends, etc).
+
+### Upgrade / Downgrade Strategy
+
+The new mode should not introduce any upgrade/downgrade problems,
+excepting that you can't downgrade or feature-disable a cluster using
+the new kube-proxy mode without switching it back to `iptables` or
+`ipvs` first.
+
+When rolling out or rolling back the feature, it should be safe to
+enable the feature gate and change the configuration at the same time,
+since nothing cares about the feature gate except for kube-proxy
+itself. Likewise, it is expected to be safe to roll out the feature in
+a live cluster, even though this will result in different proxy modes
+running on different nodes, because Kubernetes service proxying is
+defined in such a way that no node needs to be aware of the
+implementation details of the service proxy implementation on any
+other node.
+
+(However, see the notes below in [Feature Enablement and
+Rollback](#feature-enablement-and-rollback) about stale rule cleanup
+when switching modes.)
+
+### Version Skew Strategy
+
+The feature is isolated to kube-proxy and does not introduce any API
```

Comment:
Doesn't kube-proxy have versions? What happens if I use new setting with an older kube-proxy?

Reply:
That's not "Version Skew Strategy", that's "Upgrade / Downgrade Strategy".

(I don't answer that question directly there, it just says "you can't downgrade or feature-disable a cluster using the new kube-proxy mode without switching it back to `iptables` or `ipvs` first". As for what actually happens, 

`E0915 10:19:31.504389  447804 run.go:74] "command failed" err="failed validate: KubeProxyConfiguration.Mode.ProxyMode: Invalid value: \"nftables\": must be iptables, ipvs or blank (blank means the best-available proxy [currently iptables])"`

)